apt upgrade change detection

bauen1 opened the pull request:
Just a draft for adding proper change detection to `apt-get dist-upgrade` and possibly extending this to cover `apt-get upgrade` too.

Currently based on #1178 

Also there's a note on `/var/lib/apt/periodic/update-stamp` vs. `/var/lib/apt/periodic/update-success-stamp` to be addressed.

bauen1 committed:
operations.apt: Attempt to make apt-get operations detect changes

* Also move `noninteractive_apt` from operations.apt to facts.apt

Fizzadar commented on docs/using-operations.rst:
```diff
@@ -166,6 +166,11 @@ All operations return an operation meta object which provides information about
         _if=lambda: create_user.did_change() or create_otheruser.did_change(),
     )
 
+    # The functions `any_changed` and `all_changed` are provided for common use cases, e.g.
```
This is an excellent addition here!

Fizzadar commented on pyinfra/facts/apt.py:
```diff
@@ -7,6 +7,23 @@
+
+    return " ".join(args)
+
+APT_CHANGES_RE = re.compile(r"^(\d+) upgraded, (\d+) newly installed, (\d+) to remove and (\d+) not upgraded.$")
```
Interesting! This is not an approach I'd thought of using at all, I like it.

bauen1 replied:
And this line should also be present when executing e.g. `apt-get autoremove`, hence my question https://matrix.to/#/!dPvLqxqwbQZYeQrNnu:matrix.org/$ygUK_H5ACvxHxmEbKiwfSBsCQsoiTINm9WqBl5tSatE?via=beeper.com&via=matrix.org&via=t2bot.io if there was a way for operations to generate a command and then decide, based on the output, if any change was actually performed.
That could then also apply to `apt-get update` and then almost all apt operations could support proper change detection.

Fizzadar reviewed (commented):
This is really neat @bauen1!

bauen1 committed:
operations.apt: extend change detection to apt-get upgrade

bauen1 commented:
Changes for `apt-get upgrade` are untested, but you get the idea.

bauen1 committed:
pyinfra.api.host: add type hints for log, log_styled, noop

bauen1 committed:
facts.apt: add typing for SimulateOperationWillChange fact

bauen1 committed:
facts.apt: fix unexpected argument for requires_command

bauen1 committed:
pyinfra.api.deploy: add typing for the deploy decorator

bauen1 commented:
@Fizzadar The operations unit tests `FakeHost.get_fact` does not actually support positional arguments, how can I fix that ?

bauen1 committed:
facts.apt: ensure output is in english

maisim commented on pyinfra/operations/apt.py:
```diff
@@ -21,21 +27,18 @@
-    if force:
-        args.append("--force-yes")
+def _simulate_then_perform(command: str):
+    changes = host.get_fact(SimulateOperationWillChange, command)
```
Regarding your question that you asked on matrix, I think you should do:
 changes = host.get_fact(SimulateOperationWillChange, command=command)

bauen1 replied:
My problem is, that the `get_fact` method of the `FakeHost` of the framework does not support positional arguments.
But there are already other facts that use positional arguments, so this seems to be an oversight :thinking:

maisim replied:
Are you sure ? it seems to me that only the name of the Fact is a positional argument, all the others have to be named arguments, let me check that

An example where get_fact is called with positional arguments?

bauen1 replied:
Oh, that's a good point, I only though about the definiton of the `command` method using positional arguments, e.g. for `FindInFile`:

```python
def command(self, path, pattern, interpolate_variables=False):
```

But even that is called using a keyword argument:

```python
pyinfra/operations/server.py
19:from pyinfra.facts.files import Directory, FindInFile, Link
1193:        FindInFile, path=locales_definitions_file, pattern=rf"^.*{locale}[[:space:]]\+.*$"
```

So, maybe that's an oversight or some quirk that should be documented ?

In the meantime I added some rudimentary support (or rather, ignorance of) positional arguments for `get_fact`

bauen1 replied:
And the real `get_fact` method does support passing in positional arguments, but then apparently does some wild magic to "merge" positional and keyword arguments: 

https://github.com/pyinfra-dev/pyinfra/blob/3.x/pyinfra/api/facts.py#L121-L143

maisim replied:
You're right, it is possible, just not used :)
You're fix in tests/utils.py seems justified :+1:

Fizzadar replied:
Agreed this is a good change! The facts API is... old and messy, unfortunately hard to fix it without breaking it.

bauen1 replied:
> Agreed this is a good change! The facts API is... old and messy, unfortunately hard to fix it without breaking it.

I mean, if you ever do decide to change it (in a probably breaking way), one of the things that's been bothering me is, that while you usually use operations in a "namespace preserving way" i.e. `files.put` you don't do the same with Facts, i.e. `FileExists` ...

bauen1 committed:
tests: add tests for facts.apt.SimulateOperationWillChange

bauen1 committed:
WIP: make tests kind of work, but also not

bauen1 committed:
WIP: maybe make operational tests also check positional arguments

bauen1 committed:
WIP: tests: merge positional and kwargs like normal pyinfra for get_fact

bauen1 committed:
WIP: tests: complain loudly about missing fact entries, they hide bugs

bauen1 committed:
WIP: tests/operations: fix test cases using the FindInFile fact

bauen1 committed:
WIP: cleanup of c27c2bd3529ed1ec638742f44638a44fc9b4c9ec

bauen1 committed:
operations.apt: handle case where simulating operation failed

bauen1 committed:
WIP: tests/operations: make apt tests pass, but they are still wrong!

bauen1 committed:
tests/operations: misc fixes

bauen1 committed:
WIP: fix a single test to show the problem

bauen1 commented:
So, I've "fixed" the test, but this also kind of breaks existing tests, since the facts kwarg string is now different.
Actually, it doesn't even break all tests, e.g. https://github.com/pyinfra-dev/pyinfra/pull/1179/commits/c884c43370c015d17bd5fb0ed70c13dabc427550 doesn't change the outcome.

One of the reasons is, that missing fact entries currently don't raise an error or even warning, so I would propose https://github.com/pyinfra-dev/pyinfra/pull/1179/commits/b964fe557e9899674b9f66373042589fe4b20b70 or even better, make this a fatal error for the test case.


Additionally, the `idempotent` and `disable_idempotent_warning` fields should be cleaned up like https://github.com/pyinfra-dev/pyinfra/pull/1179/commits/6040d3449684dec85b9919decf2f1e3cab66dd52 - I will try to split this cleanup into it's own MR to keep this one simpler...

Fizzadar commented on tests/util.py:
```diff
@@ -194,23 +195,33 @@ def _get_fact_key(fact_cls):
         if fact is None:
             raise KeyError("Missing test fact data: {0}".format(fact_key))
+
+        # This does the same thing that pyinfra.apit.facts._handle_fact_kwargs does
```
Nice, good to have this be the same even if it is a bit weird!

Fizzadar committed:
Update Windows test

Fizzadar reviewed (approved):
This is awesome, thank you for this @bauen1 and sorry for taking so long!

I totally agree with your comment re: test keys missing. I was going to leave it as a warning but actually it seems like a good chance to fix this properly. I've pushed a quick fix for the failing windows test but will follow (today, if time allows) with updated keys in all the tests.

Fizzadar committed:
Require tests to explicitly include all fact values

This mostly impacts anywhere the operation expects the fact to return
`None`, which was the default before this change. Now requires each test
to explicitly define the fact return value or an exception will be raised.

As can be seen in the mass update of test files there were a number of
invalid test facts defined that relied on this bahavior, moving to an
explicit definition is a significant improvement.

Fizzadar commented on pyinfra/facts/apt.py:
```diff
@@ -94,3 +118,39 @@ def command(self) -> str:
+
+class SimulateOperationWillChange(FactBase[AptSimulationDict]):
+    """
+    Simulate an 'apt-get' operation and try to detect if any changes would be performed.
```
Didn't mention in my review, this is a very cool trick, love it

bauen1 replied:
Technically speaking, this primarily works around the inability to tell pyinfra if a command run actually changed something.

Ideally I'd run the `apt-get` command and additionally inform pyinfra after it run if it actually changed something.

But glad you like it, and it does improve my update scripts a bit by avoiding unnecessary reboots :)
